File: backend/services/mistake_tracker.py
from database import get_db_connection
import random

class MistakeTracker:
    def __init__(self):
        pass

    def get_stats(self, words):
        """
        Fetch stats for a list of words.
        Returns a dict mapping word -> {mistake_count, correct_count, recent_correct, recent_mistake, ratio}
        """
        if not words:
            return {}
            
        conn = get_db_connection()
        placeholders = ','.join(['?'] * len(words))
        query = f"SELECT * FROM word_stats WHERE word IN ({placeholders})"
        rows = conn.execute(query, words).fetchall()
        
        # Get recent stats for THESE words
        # (Could optimize with a WHERE IN clause in get_recent_performance, but for simplicity we fetch all or filter)
        # Let's filter the get_recent_performance logic to accept words list
        recent_stats = self.get_recent_performance(words)
        
        conn.close()

        stats = {}
        # Fill with defaults first (for words not in word_stats yet)
        for w in words:
            stats[w] = {
                'mistake_count': 0, 
                'correct_count': 0,
                'ratio': 0,
                'recent_correct': 0,
                'recent_mistake': 0,
                'recent_ratio': 0
            }

        for row in rows:
            word = row['word']
            total = row['correct_count'] + row['mistake_count']
            ratio = 0
            if total > 0:
                ratio = round(row['correct_count'] / total * 100, 1)

            recent = recent_stats.get(word, {'recent_correct': 0, 'recent_mistake': 0})
            r_total = recent['recent_correct'] + recent['recent_mistake']
            r_ratio = 0
            if r_total > 0:
                r_ratio = round(recent['recent_correct'] / r_total * 100, 1)

            stats[word] = {
                'mistake_count': row['mistake_count'],
                'correct_count': row['correct_count'],
                'ratio': ratio,
                'recent_correct': recent['recent_correct'],
                'recent_mistake': recent['recent_mistake'],
                'recent_ratio': r_ratio
            }
        
        return stats

    def get_recent_performance(self, word_list=None):
        """
        Get recent 20 exam performance for words.
        :param word_list: Optional list of words to filter. If None, returns all.
        Returns: dict mapping word -> {recent_correct, recent_mistake}
        """
        conn = get_db_connection()
        try:
            where_clause = ""
            params = []
            if word_list:
                placeholders = ','.join(['?'] * len(word_list))
                where_clause = f"WHERE word IN ({placeholders})"
                params = word_list

            # Improved query with optional filtering
            # Note: Filtering before window function is more efficient if possible, 
            # but window function needs PARTITION BY word.
            # So filtering the base table is fine.
            
            base_query = f"""
                SELECT word, is_correct, timestamp, id
                FROM exam_attempts
                {where_clause}
            """
            
            query = f"""
                SELECT word, 
                       SUM(CASE WHEN is_correct THEN 1 ELSE 0 END) as recent_correct,
                       SUM(CASE WHEN NOT is_correct THEN 1 ELSE 0 END) as recent_mistake
                FROM (
                    SELECT word, is_correct, row_number() OVER (PARTITION BY word ORDER BY timestamp DESC, id DESC) as rn
                    FROM ({base_query})
                )
                WHERE rn <= 20
                GROUP BY word
            """
            
            # Param usage: base_query needs params
            rows = conn.execute(query, params).fetchall()
            recent_stats = {}
            for row in rows:
                recent_stats[row['word']] = {
                    'recent_correct': row['recent_correct'],
                    'recent_mistake': row['recent_mistake']
                }
            return recent_stats
        except Exception as e:
            print(f"Error fetching recent stats: {e}")
            return {}
        finally:
            conn.close()

    def get_all_stats(self):
        """Fetch stats for all words in the database, including recent 20."""
        conn = get_db_connection()
        rows = conn.execute("SELECT * FROM word_stats").fetchall()
        conn.close()

        recent_stats = self.get_recent_performance()

        stats = []
        for row in rows:
            total = row['correct_count'] + row['mistake_count']
            ratio = 0
            if total > 0:
                ratio = round(row['correct_count'] / total * 100, 1)
            
            recent = recent_stats.get(row['word'], {'recent_correct': 0, 'recent_mistake': 0})
            r_total = recent['recent_correct'] + recent['recent_mistake']
            r_ratio = 0
            if r_total > 0:
                r_ratio = round(recent['recent_correct'] / r_total * 100, 1)

            stats.append({
                'word': row['word'],
                'correct_count': row['correct_count'],
                'mistake_count': row['mistake_count'],
                'ratio': ratio,
                'recent_correct': recent['recent_correct'],
                'recent_mistake': recent['recent_mistake'],
                'recent_ratio': r_ratio
            })
        return stats

    def reset_stats(self):
        """Clear all statistics from the database."""
        conn = get_db_connection()
        conn.execute("DELETE FROM word_stats")
        conn.execute("DELETE FROM exam_attempts") # Clear history too
        conn.commit()
        conn.close()

    def record_result(self, word, is_correct):
        """Update stats for a single word."""
        conn = get_db_connection()
        c = conn.cursor()
        
        # 1. Update Aggregate Stats (Lifetime)
        c.execute("SELECT 1 FROM word_stats WHERE word = ?", (word,))
        exists = c.fetchone()

        if is_correct:
            if exists:
                c.execute("UPDATE word_stats SET correct_count = correct_count + 1, last_review_date = CURRENT_TIMESTAMP WHERE word = ?", (word,))
            else:
                c.execute("INSERT INTO word_stats (word, correct_count, mistake_count) VALUES (?, 1, 0)", (word,))
        else:
            if exists:
                c.execute("UPDATE word_stats SET mistake_count = mistake_count + 1, last_review_date = CURRENT_TIMESTAMP WHERE word = ?", (word,))
            else:
                c.execute("INSERT INTO word_stats (word, correct_count, mistake_count) VALUES (?, 0, 1)", (word,))
        
        # 2. Log History for Recent Stats
        c.execute("INSERT INTO exam_attempts (word, is_correct) VALUES (?, ?)", (word, is_correct))

        conn.commit()
        conn.close()

    def generate_exam(self, all_words, num_questions=10, new_ratio=0.2, mistake_weight=2.0):
        """
        Generate an exam based on settings.
        Uses Recent 20 Mistake Count for weighting.
        :param new_ratio: Float 0.0-1.0, probability of picking a new word (0 stats).
        :param mistake_weight: Float >= 1.0, multiplier for mistake count in weighting.
        """
        if not all_words:
            return []

        # Get stats for all available words (Including recent!)
        # We need recent stats specifically for weighting mistakes
        recent_stats = self.get_recent_performance()
        
        # Also limit lifetime stats to distinguish new words (0 attempts)
        lifetime_stats = self.get_stats([w['word'] for w in all_words])

        # Split into pools
        new_words = []
        old_words = []
        old_weights = []

        for w in all_words:
            # Check if NEW (Lifetime attempts = 0)
            l_stats = lifetime_stats.get(w['word'], {'mistake_count': 0, 'correct_count': 0})
            total_attempts = l_stats['mistake_count'] + l_stats['correct_count']
            
            if total_attempts == 0:
                new_words.append(w)
            else:
                old_words.append(w)
                
                # Calculate weight based on RECENT mistakes
                # If no recent history, assume weight 1 (neutral)
                r_stats = recent_stats.get(w['word'], {'recent_correct': 0, 'recent_mistake': 0})
                recent_mistakes = r_stats['recent_mistake']
                
                # Base 1 + (Recent Mistakes * Weight)
                # This ensures we only penalize RECENT failures
                weight = 1 + (recent_mistakes * mistake_weight)
                old_weights.append(weight)

        selected_words = []
        
        # Calculate target counts
        target_new_count = int(num_questions * new_ratio)
        
        # Selection Logic
        for _ in range(num_questions):
            use_new = False
            if new_words and old_words:
                if len([x for x in selected_words if x in new_words]) < target_new_count:
                    use_new = True
                else:
                    use_new = False
            elif new_words:
                use_new = True
            else:
                use_new = False
            
            if use_new and new_words:
                selected_words.append(random.choice(new_words))
            elif old_words:
                # Weighted selection for old words
                choice = random.choices(old_words, weights=old_weights, k=1)[0]
                selected_words.append(choice)
            elif new_words:
                selected_words.append(random.choice(new_words))

        # Generate questions with options
        exam = []
        for target in selected_words:
            options = [target]
            # Pick 3 distractors
            distractors = [w for w in all_words if w['word'] != target['word']]
            if len(distractors) >= 3:
                options.extend(random.sample(distractors, 3))
            else:
                options.extend(distractors) 
            
            random.shuffle(options)
            
            exam.append({
                'word': target['word'],
                'correct_translation': target['translation'],
                'options': [{'word': o['word'], 'translation': o['translation']} for o in options]
            })

        return exam

File: frontend/src/services/MistakeTracker.js
const STATS_KEY = 'vocab_master_stats';
const HISTORY_KEY = 'vocab_master_history';

const MistakeTracker = {
    getStats(words) {
        const stats = this._getAllStats();
        const history = this._getHistory();

        // Calculate recent stats for requested words
        const recentStats = {};
        const recentWindow = 20;

        // Group history by word
        const wordHistory = {};
        history.forEach(entry => {
            if (!wordHistory[entry.word]) {
                wordHistory[entry.word] = [];
            }
            wordHistory[entry.word].push(entry);
        });

        // Compute recent performance
        words.forEach(word => {
            const entries = wordHistory[word] || [];
            // Sort by timestamp desc (newest first)
            entries.sort((a, b) => b.timestamp - a.timestamp);

            const recentEntries = entries.slice(0, recentWindow);

            let recentCorrect = 0;
            let recentMistake = 0;

            recentEntries.forEach(e => {
                if (e.isCorrect) recentCorrect++;
                else recentMistake++;
            });

            recentStats[word] = { recentCorrect, recentMistake };
        });

        // Assemble final stats object
        const result = {};
        words.forEach(word => {
            const s = stats[word] || { mistakeCount: 0, correctCount: 0 };
            const r = recentStats[word];

            const total = s.correctCount + s.mistakeCount;
            const ratio = total > 0 ? Math.round((s.correctCount / total) * 1000) / 10 : 0;

            const rTotal = r.recentCorrect + r.recentMistake;
            const rRatio = rTotal > 0 ? Math.round((r.recentCorrect / rTotal) * 1000) / 10 : 0;

            result[word] = {
                mistake_count: s.mistakeCount,
                correct_count: s.correctCount,
                ratio: ratio,
                recent_correct: r.recentCorrect,
                recent_mistake: r.recentMistake,
                recent_ratio: rRatio
            };
        });

        return result;
    },

    getAllStats() {
        const stats = this._getAllStats();
        const words = Object.keys(stats);
        if (words.length === 0) return [];

        const statsMap = this.getStats(words);
        return Object.keys(statsMap).map(word => ({
            word,
            ...statsMap[word]
        }));
    },

    recordResult(word, isCorrect) {
        const stats = this._getAllStats();

        if (!stats[word]) {
            stats[word] = { correctCount: 0, mistakeCount: 0 };
        }

        if (isCorrect) {
            stats[word].correctCount++;
        } else {
            stats[word].mistakeCount++;
        }
        stats[word].lastReview = Date.now();

        this._saveStats(stats);
        this._addToHistory(word, isCorrect);
    },

    generateExam(allWords, settings) {
        const { numQuestions = 10, newRatio = 20, mistakeWeight = 5.0 } = settings;
        const newRatioDecimal = newRatio / 100.0;

        if (!allWords || allWords.length === 0) return [];

        // Get stats for all words to distinguish new vs old
        const wordList = allWords.map(w => w.word);
        const stats = this.getStats(wordList);

        const newWords = [];
        const oldWords = [];
        const oldWeights = [];

        allWords.forEach(w => {
            const s = stats[w.word];
            const totalAttempts = s ? (s.mistake_count + s.correct_count) : 0;

            if (totalAttempts === 0) {
                newWords.push(w);
            } else {
                oldWords.push(w);
                // Weight calculation: 1 + (Recent Mistakes * Weight)
                // Access recent_mistake directly from stats
                const recentMistakes = s ? s.recent_mistake : 0;
                const weight = 1 + (recentMistakes * mistakeWeight);
                oldWeights.push(weight);
            }
        });

        const selectedWords = [];
        const targetNewCount = Math.floor(numQuestions * newRatioDecimal);

        // Selection Logic
        for (let i = 0; i < numQuestions; i++) {
            let useNew = false;

            // Determine if we should pick a new word
            const currentNewCount = selectedWords.filter(w => newWords.includes(w)).length;

            if (newWords.length > 0 && oldWords.length > 0) {
                if (currentNewCount < targetNewCount) {
                    useNew = true;
                } else {
                    useNew = false;
                }
            } else if (newWords.length > 0) {
                useNew = true;
            } else {
                useNew = false;
            }

            // Pick word
            if (useNew && newWords.length > 0) {
                const randomIndex = Math.floor(Math.random() * newWords.length);
                selectedWords.push(newWords[randomIndex]);
            } else if (oldWords.length > 0) {
                // Weighted selection
                const selected = this._weightedRandom(oldWords, oldWeights);
                selectedWords.push(selected);
            } else if (newWords.length > 0) {
                // Fallback if no old words
                const randomIndex = Math.floor(Math.random() * newWords.length);
                selectedWords.push(newWords[randomIndex]);
            }
        }

        // Generate questions
        return selectedWords.map(target => {
            const formatSetting = settings.examFormat || 'standard';
            let useAtomic = formatSetting === 'atomic';
            if (formatSetting === 'mixed') {
                useAtomic = Math.random() > 0.5;
            }

            if (useAtomic) {
                // Determine base value for distractors
                const baseVal = parseInt(target.phonetic);
                let isNumeric = !isNaN(baseVal);
                let optionsData = [target];

                if (isNumeric) {
                    // Generate numeric distractors
                    const generated = new Set();
                    generated.add(baseVal);

                    while (generated.size < 4) {
                        const offset = Math.floor(Math.random() * 7) - 3; // -3 to +3
                        const distractorVal = baseVal + offset;
                        if (!generated.has(distractorVal)) {
                            generated.add(distractorVal);
                            const formattedVal = distractorVal > 0 ? `+${distractorVal}` : `${distractorVal}`;
                            optionsData.push({
                                word: `dummy_${formattedVal}`,
                                translation: target.zh, // Unused for display in atomic answer, but keep valid obj
                                phonetic: formattedVal
                            });
                        }
                    }
                } else {
                    // Fallback to purely random phonetic values from other words if it's not a number
                    const distractors = allWords.filter(w => w.word !== target.word && w.phonetic);
                    if (distractors.length >= 3) {
                        const shuffled = [...distractors].sort(() => 0.5 - Math.random());
                        optionsData = optionsData.concat(shuffled.slice(0, 3));
                    } else {
                        optionsData = optionsData.concat(distractors);
                    }
                }

                // Shuffle options
                optionsData.sort(() => 0.5 - Math.random());

                return {
                    isAtomic: true,
                    word: `${target.word} (${target.zh})`,
                    correct_translation: target.phonetic, // Only the phonetic part is the answer
                    target_word: target.word, // For tracking stats correctly
                    options: optionsData.map(o => ({
                        word: o.word,
                        isAtomicOption: true,
                        translation: o.phonetic // Display phonetic as the option text
                    }))
                };
            }

            // Standard Generation Fallback
            const distractors = allWords.filter(w => w.word !== target.word);
            let options = [target];

            if (distractors.length >= 3) {
                // Shuffle distractors and pick 3
                const shuffled = [...distractors].sort(() => 0.5 - Math.random());
                options = options.concat(shuffled.slice(0, 3));
            } else {
                options = options.concat(distractors);
            }

            // Shuffle options
            options.sort(() => 0.5 - Math.random());

            return {
                isAtomic: false,
                target_word: target.word,
                word: target.word,
                correct_translation: target.translation,
                options: options.map(o => ({
                    word: o.word,
                    translation: o.translation,
                    zh: o.zh,
                    phonetic: o.phonetic
                }))
            };
        });
    },

    resetStats() {
        localStorage.removeItem(STATS_KEY);
        localStorage.removeItem(HISTORY_KEY);
    },

    // Internal Helpers
    _getAllStats() {
        const stored = localStorage.getItem(STATS_KEY);
        return stored ? JSON.parse(stored) : {};
    },

    _saveStats(stats) {
        localStorage.setItem(STATS_KEY, JSON.stringify(stats));
    },

    _getHistory() {
        const stored = localStorage.getItem(HISTORY_KEY);
        return stored ? JSON.parse(stored) : [];
    },

    _addToHistory(word, isCorrect) {
        const history = this._getHistory();
        history.push({
            word,
            isCorrect,
            timestamp: Date.now(),
            id: generateId()
        });
        // Optional: trim history if it gets too large? 
        // For now, keep it simple.
        localStorage.setItem(HISTORY_KEY, JSON.stringify(history));
    },

    _weightedRandom(items, weights) {
        const totalWeight = weights.reduce((sum, w) => sum + w, 0);
        let random = Math.random() * totalWeight;

        for (let i = 0; i < items.length; i++) {
            random -= weights[i];
            if (random <= 0) {
                return items[i];
            }
        }
        return items[items.length - 1]; // Fallback
    }
};

function generateId() {
    return Math.random().toString(36).substr(2, 9);
}

export default MistakeTracker;
